src/nnet3/nnet-discriminative-training.cc

```diff
@@ -0,0 +1,289 @@
+    
+    if (use_xent) {
+      // this block computes the cross-entropy objective.
+      const CuMatrixBase<BaseFloat> &xent_output = computer->GetOutput(xent_name);
+      // at this point, xent_deriv is posteriors derived from the numerator
+      // computation.  note, xent_objf has a factor of '.supervision.weight'
+      BaseFloat xent_objf = TraceMatMat(xent_output, xent_deriv, kTrans);
+      if (xent_objf != xent_objf) {
+        BaseFloat default_objf = -10;
+        xent_objf = default_objf;
+      }
+
+      objf_info_[xent_name].UpdateStats(xent_name, "xent",
+                                        opts_.nnet_config.print_interval,
+                                        num_minibatches_processed_,
+                                        stats.TotalT(), xent_objf);
+    }
+
+    if (opts_.apply_deriv_weights && sup.deriv_weights.Dim() != 0) {
+      CuVector<BaseFloat> cu_deriv_weights(sup.deriv_weights);
+      nnet_output_deriv.MulRowsVec(cu_deriv_weights);
+      if (use_xent)
+        xent_deriv.MulRowsVec(cu_deriv_weights);
+    }
+
+    computer->AcceptOutputDeriv(sup.name, &nnet_output_deriv);
+
+    objf_info_[sup.name].UpdateStats(sup.name, opts_.discriminative_training_config.criterion,
+                                     opts_.nnet_config.print_interval,
+                                     num_minibatches_processed_++,
+                                     stats);
+    
+    if (use_xent) {
+      xent_deriv.Scale(opts_.discriminative_training_config.xent_regularize);
+      computer->AcceptOutputDeriv(xent_name, &xent_deriv);
+    }
+  }
+}
+
+
+bool NnetDiscriminativeTrainer::PrintTotalStats() const {
+  unordered_map<std::string, DiscriminativeObjectiveFunctionInfo>::const_iterator
+      iter = objf_info_.begin(),
+      end = objf_info_.end();
+  bool ans = false;
+  for (; iter != end; ++iter) {
+    const std::string &name = iter->first;
+    const DiscriminativeObjectiveFunctionInfo &info = iter->second;
+    bool ret = info.PrintTotalStats(name, opts_.discriminative_training_config.criterion);
+    ans = ans || ret;
+  }
+
+  return ans;
+}
+
+
+void DiscriminativeObjectiveFunctionInfo::UpdateStats(
+    const std::string &output_name,
+    const std::string &criterion,
+    int32 minibatches_per_phase,
+    int32 minibatch_counter,
+    BaseFloat this_minibatch_weight,
+    BaseFloat this_minibatch_tot_objf,
+    BaseFloat this_minibatch_tot_aux_objf) {
+  int32 phase = minibatch_counter / minibatches_per_phase;
+  if (phase != current_phase) {
+    KALDI_ASSERT(phase == current_phase + 1); // or doesn't really make sense.
+    PrintStatsForThisPhase(output_name, criterion, minibatches_per_phase);
+    current_phase = phase;
+    stats_this_phase.Reset();
+    tot_aux_objf_this_phase = 0.0;
+  }
+  stats_this_phase.tot_t_weighted += this_minibatch_weight;
+  stats_this_phase.tot_objf += this_minibatch_tot_objf;
+  tot_aux_objf_this_phase += this_minibatch_tot_aux_objf;
+
+  stats.Add(stats_this_phase);
```

Comment:
this looks like a bug- you seem to be adding stats_this_phase to stats each time, so you are counting the same minibatch multiple times.